random dag

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142180

Utils for creating random dags and generating code for nn modules based on such dags.

In particular, this was used to do fuzz testing for unflatten, where the random dags instructed the generation of calls, const accesses, and buffer mutations in a system of nn modules.

Example of generated test:
```python
    def test_unflatten_random_dag_const_preserving_3(self):
        class N2(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.const = torch.ones(1)

            def forward(self, x):
                return x + 1

        class N1(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.const = torch.ones(1)
                self.n2 = N2()

            def forward(self, x):
                x = x + self.n2.const
                x = self.n2(x + 1)
                return x + 1

        class N0(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.const = torch.ones(1)
                self.n1 = N1()

            def forward(self, x):
                x = x + self.n1.n2.const
                x = self.n1(x + 1)
                x = self.n1.n2(x + 1)
                return x + 1

        inp = (torch.ones(1),)
        eager = N0()(*inp)
        ep = torch.export.export(
            N0(),
            inp,
            strict=False,
            preserve_module_call_signature=(
                "n1",
                "n1.n2",
            ),
        )
        epm = ep.module()
        ufm = torch.export.unflatten(ep)
        assert torch.allclose(epm(*inp), eager)
        assert torch.allclose(ufm(*inp), eager)
```

Differential Revision: [D66838348](https://our.internmc.facebook.com/intern/diff/D66838348/)

avikchaudhuri committed:
random dag

Utils for creating random dags and generating code for nn modules based on such dags.

In particular, this was used to do fuzz testing for unflatten, where the random dags instructed the generation of calls, const accesses, and buffer mutations in a system of nn modules.

Differential Revision: [D66838348](https://our.internmc.facebook.com/intern/diff/D66838348/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "random dag"


Utils for creating random dags and generating code for nn modules based on such dags.

In particular, this was used to do fuzz testing for unflatten, where the random dags instructed the generation of calls, const accesses, and buffer mutations in a system of nn modules.

Differential Revision: [D66838348](https://our.internmc.facebook.com/intern/diff/D66838348/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "random dag"


Utils for creating random dags and generating code for nn modules based on such dags.

In particular, this was used to do fuzz testing for unflatten, where the random dags instructed the generation of calls, const accesses, and buffer mutations in a system of nn modules.

Differential Revision: [D66838348](https://our.internmc.facebook.com/intern/diff/D66838348/)

[ghstack-poisoned]

ydwu4 commented on test/export/random_dag.py:
```diff
@@ -0,0 +1,273 @@
+            if random.choice([True, False]):
+                edges[i].append(j)
+
+    print(edges)
```
nit: can remove print?

ydwu4 commented on test/export/random_dag.py:
```diff
@@ -0,0 +1,273 @@
+# Owner(s): ["oncall: export"]
+
+
+def random_dag(n):
```
Some type annotation maybe?

ydwu4 reviewed (approved):
Looks good. Can we have some sample tests generated in the PR description?

avikchaudhuri committed:
Update on "random dag"


Utils for creating random dags and generating code for nn modules based on such dags.

In particular, this was used to do fuzz testing for unflatten, where the random dags instructed the generation of calls, const accesses, and buffer mutations in a system of nn modules.

Differential Revision: [D66838348](https://our.internmc.facebook.com/intern/diff/D66838348/)

[ghstack-poisoned]
